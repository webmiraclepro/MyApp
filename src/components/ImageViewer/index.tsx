import React, { useRef, useCallback } from 'react'
import {
  StyleSheet,
  ImageProps,
  Dimensions,
  View,
  Image
} from 'react-native'
import ImageZoom, { ImageZoomProps, IOnClick } from 'react-native-image-pan-zoom'

type ImageZoomWithoutUsedProps = Omit<ImageZoomProps, keyof {
  cropWidth: ImageZoomProps['cropWidth'],
  cropHeight: ImageZoomProps['cropHeight'],
  imageWidth: ImageZoomProps['imageWidth'],
  imageHeight: ImageZoomProps['imageHeight'],
  minScale: ImageZoomProps['minScale'],
  onMove: ImageZoomProps['onMove'],
  onStartShouldSetPanResponder: ImageZoomProps['onStartShouldSetPanResponder']
}>

export type ImageViewerProps = ImageZoomWithoutUsedProps & {
  source: ImageProps['source'],
  width?: number,
  height?: number,
  onZoom?: (scale: number) => void
}

const ImageViewer: React.FC<ImageViewerProps> = ({
  source,
  width = Dimensions.get('window').width,
  height = Dimensions.get('window').height,
  onZoom,
  ...rest
}) => {
  const scaleValue = useRef(1)
  const imageZoomRef = useRef<ImageZoom>(null)
  const handlDoubleClick = useCallback(({ pageX, pageY }: IOnClick) =>  {
    if (imageZoomRef.current && scaleValue.current === 1) {
      imageZoomRef.current.centerOn({
        x: pageX,
        y: pageY,
        scale: 3,
        duration: 300
      })
    }
  }, [scaleValue, imageZoomRef])
  return (
    <ImageZoom
      ref={imageZoomRef}
      cropWidth={Dimensions.get('window').width}
      cropHeight={Dimensions.get('window').height}
      imageWidth={width}
      imageHeight={height}
      minScale={1}
      maxScale={7}
      enableDoubleClickZoom
      enableCenterFocus={false}
      pinchToZoom
      useNativeDriver
      panToMove
      enableSwipeDown={false}
      {...rest}
      onMove={({ scale }) => {
        scaleValue.current = scale
        onZoom && onZoom(scale)
      }}
      onDoubleClick={(e) => handlDoubleClick(e)}
      onStartShouldSetPanResponder={(e) => {
        return e.nativeEvent.touches.length === 2 || scaleValue.current > 1
      }}
    >
      <View
        style={styles.container}
        onStartShouldSetResponder={(e) => {
          return e.nativeEvent.touches.length < 2 && scaleValue.current <= 1
        }}
      >
        <Image
          width={width}
          height={height}
          source={source}
          resizeMode='contain'
          style={styles.container}
        />
      </View>
    </ImageZoom>
  )
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    flex: 1
  }
})

export default ImageViewer
