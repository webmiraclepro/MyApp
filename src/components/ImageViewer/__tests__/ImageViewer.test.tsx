import 'react-native'
import React from 'react'
import { mount, shallow } from 'enzyme'

import ImageViewer, { ImageViewerProps } from '..'

describe('<ImageViewer />', () => {
  const defaultProps: ImageViewerProps = {
    source: { uri: 'something' }
  }

  it('should render', () => {
    const rendered = mount(<ImageViewer {...defaultProps} />)
    expect(rendered).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(shallow(<ImageViewer {...defaultProps} />)).toMatchSnapshot()
  })
})
