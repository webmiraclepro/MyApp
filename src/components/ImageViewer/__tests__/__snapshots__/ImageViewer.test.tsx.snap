// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageViewer /> renders correctly 1`] = `
<ImageViewer
  clickDistance={10}
  cropHeight={1334}
  cropWidth={750}
  doubleClickInterval={175}
  enableCenterFocus={false}
  enableDoubleClickZoom={true}
  enableSwipeDown={false}
  horizontalOuterRangeOffset={[Function]}
  imageHeight={1334}
  imageWidth={750}
  layoutChange={[Function]}
  longPressTime={800}
  maxOverflow={100}
  maxScale={7}
  minScale={1}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onDragLeft={[Function]}
  onLongPress={[Function]}
  onMove={[Function]}
  onPanResponderTerminationRequest={[Function]}
  onStartShouldSetPanResponder={[Function]}
  onSwipeDown={[Function]}
  panToMove={true}
  pinchToZoom={true}
  responderRelease={[Function]}
  style={Object {}}
  swipeDownThreshold={230}
  useHardwareTextureAndroid={true}
  useNativeDriver={true}
>
  <View
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Image
      height={1334}
      resizeMode="contain"
      source={
        Object {
          "uri": "something",
        }
      }
      style={
        Object {
          "flex": 1,
          "height": "100%",
          "width": "100%",
        }
      }
      width={750}
    />
  </View>
</ImageViewer>
`;
